Incorrect type definition for React elements

- [ x] I tried using the latest `xxxx/xxxx.d.ts` file in this repo and had problems.
- [ x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x ] I want to talk about `xxxx/xxxx.d.ts`.
  - The authors of that type definition are cc/ @....


According to the [official site](https://facebook.github.io/react/docs/conditional-rendering.html#preventing-component-from-rendering), React components may return `null` to prevent any rendering. But due to a likely bug in the definition files, I get a type error when trying to use a component that possibly returns `null`.  Here is the compiler error:

```
    (54,3): error TS2345: Argument of type '(props: PaginationProps) => Element | null' is not assignable to parameter of type 'ComponentClass<PaginationProps & {}> | StatelessComponent<PaginationProps & {}>'.
      Type '(props: PaginationProps) => Element | null' is not assignable to type 'StatelessComponent<PaginationProps & {}>'.
        Type 'Element | null' is not assignable to type 'ReactElement<any>'.
          Type 'null' is not assignable to type 'ReactElement<any>'.
```

I think the fix is to make `ReactElement` nullable

--- a/types/react-native/index.d.ts
+++ b/types/react-native/index.d.ts
@@ -1,6 +1,8 @@
 // Type definitions for react-native 0.44
 // Project: https://github.com/facebook/react-native
-// Definitions by: Eloy Durán <https://github.com/alloy>, Fedor Nezhivoi <https://github.com/gyzerok>, HuHuanming <https://github.com/huhuanming>, Jeremi Stadler <https://github.com/jeremistadler>, Kyle Roach <https://github.com/iRoachie>
+// Definitions by: Eloy Durán <https://github.com/alloy>, Fedor Nezhivoi <https://github.com/gyzerok>,
+//  HuHuanming <https://github.com/huhuanming>, Jeremi Stadler <https://github.com/jeremistadler>,
+//  Kyle Roach <https://github.com/iRoachie>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.3
 

--- a/types/react-router-native/index.d.ts
+++ b/types/react-router-native/index.d.ts
@@ -2,7 +2,7 @@
 // Project: https://github.com/ReactTraining/react-router-native
 // Definitions by: Eduard Zintz <https://github.com/ezintz>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.2
+// TypeScript Version: 2.3
 
 export {
   match,
@@ -23,25 +23,25 @@ import * as React from 'react';
 import * as H from 'history';
 
 export interface AndroidBackButtonProps {
-  children: Array<React.Component<void, void>>;
+  children: Array<React.Component<{}, {}>>;
 }
 
-export class AndroidBackButton extends React.Component<AndroidBackButtonProps, void> {}
+export class AndroidBackButton extends React.Component<AndroidBackButtonProps, {}> {}
 
 export class DeepLinking extends React.Component<{}, {}> {}
 
 export interface LinkProps {
-  component?: React.Component<void, void>;
+  component?: React.Component<{}, {}>;
   replace?: boolean;
   style?: any;
   to: H.LocationDescriptor;
 }
 
-export class Link extends React.Component<LinkProps, void> {}
+export class Link extends React.Component<LinkProps, {}> {}
 
 export interface NativeRouterProps {
   getUserConfirmation?: Function;
   keyLength?: number;
 }
 
-export class NativeRouter extends React.Component<NativeRouterProps, void> {}
+export class NativeRouter extends React.Component<NativeRouterProps, {}> {}

--- a/types/react/index.d.ts
+++ b/types/react/index.d.ts
@@ -227,7 +227,7 @@ declare namespace React {
 
     type SFC<P = {}> = StatelessComponent<P>;
     interface StatelessComponent<P = {}> {
-        (props: P & { children?: ReactNode }, context?: any): ReactElement<any>;
+        (props: P & { children?: ReactNode }, context?: any): ReactElement<any> | null;
         propTypes?: ValidationMap<P>;
         contextTypes?: ValidationMap<any>;
         defaultProps?: Partial<P>;

--- a/types/react/test/index.ts
+++ b/types/react/test/index.ts
@@ -138,7 +138,7 @@ interface SCProps {
 }
 
 function StatelessComponent(props: SCProps) {
-    return React.DOM.div(null, props.foo);
+    return props.foo ? React.DOM.div(null, props.foo) : null;
 };
 namespace StatelessComponent {
     export var displayName = "StatelessComponent";
@@ -155,7 +155,8 @@ StatelessComponent2.defaultProps = {
 
 var StatelessComponent3: React.SFC<SCProps> =
     // allows usage of props.children
-    props => React.DOM.div(null, props.foo, props.children);
+    // allows null return
+    props => props.foo ? React.DOM.div(null, props.foo, props.children) : null;
 
 // React.createFactory
 var factory: React.CFactory<Props, ModernComponent> =

